From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Herman van Hazendonk <github.com@herrie.org>
Date: Tue, 20 Oct 2026 16:22:36 +0200
Subject: [PATCH] Only require an email address for password login

discord_login refused any username without an '@' before looking at
how the account signs in. That is only needed for the email+password
path that posts to /auth/login. Accounts that use a stored token, or
QR code login with an empty password, do not need an email address as
username, but were rejected before the QR code was ever shown.

Move the check into the password branch.

Signed-off-by: Herman van Hazendonk <github.com@herrie.org>
Upstream-Status: Pending
---
 libdiscord.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/libdiscord.c b/libdiscord.c
index d9fd21e..3bbfe4f 100644
--- a/libdiscord.c
+++ b/libdiscord.c
@@ -6391,10 +6391,6 @@ discord_login(PurpleAccount *account)
 	PurpleConnection *pc = purple_account_get_connection(account);
 	PurpleConnectionFlags pc_flags;
 
-	if (!strchr(purple_account_get_username(account), '@')) {
-		purple_connection_error(pc, PURPLE_CONNECTION_ERROR_INVALID_USERNAME, _("Username needs to be an email address"));
-		return;
-	}
 
 	pc_flags = purple_connection_get_flags(pc);
 	pc_flags |= PURPLE_CONNECTION_FLAG_HTML;
@@ -6448,6 +6444,12 @@ discord_login(PurpleAccount *account)
 		discord_start_socket(da);
 
 	} else if (account_password && *account_password) {
+		/* The email+password path, and the only one that needs an address. */
+		if (!strchr(purple_account_get_username(account), '@')) {
+			purple_connection_error(pc, PURPLE_CONNECTION_ERROR_INVALID_USERNAME, _("Username needs to be an email address"));
+			return;
+		}
+
 		JsonObject *data = json_object_new();
 		gchar *str;
 
-- 
2.53.0

